File: main_pipeline.py
import cv2
import numpy as np
import json
import os
from datetime import datetime
from enum import Enum
from mediapipe.tasks import python as mp_python
from mediapipe.tasks.python import vision
from mediapipe import Image, ImageFormat
from face_recognition_engine import FaceRecognitionEngine
from liveness_detection import LivenessDetector, LivenessState

class PipelineState(Enum):
    IDLE = 0
    LIVENESS_CHECK = 1
    RECOGNIZING = 2
    SUCCESS = 3
    FAILED = -1

class AttendancePipeline:
    def __init__(self):
        print("Initializing pipeline...")
        self.face_engine = FaceRecognitionEngine()
        self.liveness = LivenessDetector()
        self.state = PipelineState.IDLE
        self.result = None
        self.result_timer = 0
        self.RESULT_DISPLAY_FRAMES = 90  # show result for ~3 seconds

        # Attendance log
        self.LOG_PATH = "attendance_log.json"
        self.attendance_log = self.load_log()
        print("✅ Pipeline ready!")

    def load_log(self):
        if os.path.exists(self.LOG_PATH):
            with open(self.LOG_PATH, 'r') as f:
                return json.load(f)
        return []

    def save_log(self):
        with open(self.LOG_PATH, 'w') as f:
            json.dump(self.attendance_log, f, indent=2)

    def log_attendance(self, person_id, score):
        entry = {
            "person_id": person_id,
            "timestamp": datetime.now().isoformat(),
            "confidence": round(score, 4),
            "synced": False  # AWS sync flag
        }
        self.attendance_log.append(entry)
        self.save_log()
        print(f"📝 Attendance logged: {person_id} at {entry['timestamp']}")

    def start(self):
        self.state = PipelineState.LIVENESS_CHECK
        self.liveness.start()
        self.result = None
        self.result_timer = 0

    def process_frame(self, frame):
        """
        Returns: (display_frame, status_message, pipeline_state)
        """
        display = frame.copy()

        if self.state == PipelineState.IDLE:
            self._draw_text(display, "Press SPACE to start", (30, 50), (255, 255, 255))
            return display, "idle", self.state

        elif self.state == PipelineState.LIVENESS_CHECK:
            liveness_state, message, landmarks = self.liveness.process_frame(frame)

            # Draw liveness UI
            color = (0, 165, 255)
            if liveness_state == LivenessState.FAILED:
                color = (0, 0, 255)
                self.state = PipelineState.FAILED
                self.result = "Liveness check failed!"
                self.result_timer = self.RESULT_DISPLAY_FRAMES
            elif liveness_state == LivenessState.VERIFIED:
                # Liveness passed → now recognize
                self.state = PipelineState.RECOGNIZING
                color = (0, 255, 0)

            self._draw_text(display, message, (30, 50), color)
            self._draw_text(display, f"Step: {liveness_state.name}", (30, 90), (255, 255, 255))
            self._draw_progress(display, liveness_state)

        elif self.state == PipelineState.RECOGNIZING:
            self._draw_text(display, "Recognizing...", (30, 50), (255, 255, 0))
            person_id, score, status = self.face_engine.recognize_face(frame)

            if person_id:
                self.log_attendance(person_id, score)
                self.result = f"Welcome, {person_id}! ({score:.0%})"
                self.state = PipelineState.SUCCESS
            else:
                self.result = f"Unknown person (score: {score:.2f})"
                self.state = PipelineState.FAILED

            self.result_timer = self.RESULT_DISPLAY_FRAMES

        elif self.state == PipelineState.SUCCESS:
            self._draw_text(display, self.result, (30, 50), (0, 255, 0))
            self._draw_text(display, "✅ Attendance logged!", (30, 90), (0, 255, 0))
            self.result_timer -= 1
            if self.result_timer <= 0:
                self.state = PipelineState.IDLE

        elif self.state == PipelineState.FAILED:
            self._draw_text(display, self.result or "Failed!", (30, 50), (0, 0, 255))
            self.result_timer -= 1
            if self.result_timer <= 0:
                self.state = PipelineState.IDLE

        return display, self.result, self.state

    def _draw_text(self, frame, text, pos, color):
        cv2.putText(frame, text, pos,
                    cv2.FONT_HERSHEY_SIMPLEX, 0.8, color, 2)

    def _draw_progress(self, frame, liveness_state):
        color = (0, 255, 0) if liveness_state == LivenessState.VERIFIED else \
            (0, 0, 255) if liveness_state == LivenessState.FAILED else \
                (0, 165, 255)
        cv2.putText(frame, f"Liveness: {liveness_state.name}", (30, 130),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.6, color, 2)

if __name__ == "__main__":
    pipeline = AttendancePipeline()
    cap = cv2.VideoCapture(0)

    print("\n=== ATTENDANCE PIPELINE ===")
    print("SPACE → Start liveness check")
    print("R     → Register new face")
    print("Q     → Quit\n")

    register_mode = False
    register_name = ""

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        frame = cv2.flip(frame, 1)  # Mirror

        if register_mode:
            cv2.putText(frame, f"REGISTER MODE: {register_name}", (30, 50),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 255, 255), 2)
            cv2.putText(frame, "Press ENTER to capture", (30, 90),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 1)
        else:
            display, result, state = pipeline.process_frame(frame)
            frame = display

        cv2.imshow('Attendance System', frame)

        key = cv2.waitKey(1) & 0xFF

        if key == ord('q'):
            break
        elif key == ord(' ') and not register_mode:
            pipeline.start()
        elif key == ord('r') and not register_mode:
            register_name = input("Enter person name/ID: ")
            register_mode = True
        elif key == 13 and register_mode:  # ENTER
            success, msg = pipeline.face_engine.register_face(frame, register_name)
            print(msg)
            register_mode = False
            register_name = ""

    cap.release()
    cv2.destroyAllWindows()

    print(f"\n📋 Total attendance records: {len(pipeline.attendance_log)}")

File: cpp/main_pipeline.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>
#include "storage.hpp"
#include "face_recognition.hpp"
#include "liveness_detector.hpp"

enum class PipelineState {
    IDLE = 0,
    PROCESSING = 1, // Combined active analysis state
    SUCCESS = 3,
    FAILED = -1
};

class AttendancePipeline {
public:
    AttendancePipeline()
        : face_engine_(), liveness_(), storage_(),
          state_(PipelineState::IDLE), result_timer_(0),
          RESULT_DISPLAY_FRAMES_(90), AUTO_RESTART_FRAMES_(60),
          liveness_passed_(false), face_recognized_(false) {
        std::cout << "Pipeline ready!" << std::endl;
    }

    void start() {
        state_ = PipelineState::PROCESSING;
        liveness_.start();
        result_ = "";
        result_timer_ = 0;
        liveness_passed_ = false;
        face_recognized_ = false;
        recognized_person_ = "";
        last_recognition_score_ = 0.0;
        std::cout << ">> Pipeline started (Simultaneous Liveness + Recognition tracking active)" << std::endl;
    }

    std::tuple<cv::Mat, std::string, PipelineState>
    process_frame(const cv::Mat& frame) {
        cv::Mat display = frame.clone();

        if (state_ == PipelineState::IDLE) {
            draw_text(display, "Press SPACE to start", {30, 50}, {255, 255, 255});
            return {display, "idle", state_};
        }

        if (state_ == PipelineState::PROCESSING) {
            // 1. Run Liveness Detection Check
            auto [ls, liveness_msg, landmarks] = liveness_.process_frame(frame);

            // Update liveness status if it hasn't passed yet
            if (!liveness_passed_) {
                if (ls == LivenessState::VERIFIED) {
                    liveness_passed_ = true;
                    std::cout << ">> Liveness verified via active frame analysis!" << std::endl;
                } else if (ls == LivenessState::FAILED) {
                    std::cout << ">> Liveness processing reached absolute timeout." << std::endl;
                    state_ = PipelineState::FAILED;
                    result_ = "Liveness check failed!";
                    result_timer_ = RESULT_DISPLAY_FRAMES_;
                    return {display, result_, state_};
                }
            }

            // 2. Run Face Recognition (Parallel execution pass on the same frame matrix)
            if (!face_recognized_) {
                auto [person_id, score, status] = face_engine_.recognize_face(frame, 0.32); // Lowered threshold filter
                last_recognition_score_ = score;
                
                if (!person_id.empty()) {
                    face_recognized_ = true;
                    recognized_person_ = person_id;
                    std::cout << ">> Face matched successfully: " << person_id << " (score: " << score << ")" << std::endl;
                }
            }

            // 3. Render real-time overlay graphics to screen layout
            cv::Scalar liveness_color = liveness_passed_ ? cv::Scalar(0, 255, 0) : cv::Scalar(0, 165, 255);
            std::string liveness_ui_msg = liveness_passed_ ? "Liveness: ALIVE!" : liveness_msg;
            draw_text(display, liveness_ui_msg, {30, 50}, liveness_color);

            cv::Scalar rec_color = face_recognized_ ? cv::Scalar(0, 255, 0) : cv::Scalar(0, 255, 255);
            std::string rec_ui_msg = face_recognized_ ? "Identity: " + recognized_person_ : "Identity: Scanning...";
            draw_text(display, rec_ui_msg, {30, 90}, rec_color);

            // 4. State transition handling once both criteria match
            if (liveness_passed_ && face_recognized_) {
                storage_.log(recognized_person_, last_recognition_score_);
                result_ = "Welcome, " + recognized_person_ + "!";
                state_ = PipelineState::SUCCESS;
                result_timer_ = RESULT_DISPLAY_FRAMES_;
            }

            return {display, result_, state_};
        }

        if (state_ == PipelineState::SUCCESS) {
            draw_text(display, result_, {30, 50}, {0, 255, 0});
            draw_text(display, "Attendance logged!", {30, 90}, {0, 255, 0});
            result_timer_--;
            if (result_timer_ <= AUTO_RESTART_FRAMES_ && result_timer_ > 0) {
                draw_text(display, "Auto-restarting...", {30, 130}, {255, 255, 255});
            }
            if (result_timer_ <= 0) {
                start();
            }
            return {display, result_, state_};
        }

        if (state_ == PipelineState::FAILED) {
            draw_text(display, result_.empty() ? "Failed!" : result_, {30, 50}, {0, 0, 255});
            draw_text(display, "SPACE to retry, auto-restarting...", {30, 90}, {255, 255, 255});
            result_timer_--;
            if (result_timer_ <= 0) {
                start();
            }
            return {display, result_, state_};
        }

        return {display, result_, state_};
    }

    void draw_text(cv::Mat& frame, const std::string& text, cv::Point pos, cv::Scalar color) {
        cv::putText(frame, text, pos, cv::FONT_HERSHEY_SIMPLEX, 0.8, color, 2);
    }

    AttendanceStorage& get_storage() { return storage_; }
    std::pair<bool, std::string> register_face(const cv::Mat& image, const std::string& person_id) {
        return face_engine_.register_face(image, person_id);
    }

private:
    FaceRecognitionEngine face_engine_;
    LivenessDetector liveness_;
    AttendanceStorage storage_;
    PipelineState state_;
    std::string result_;
    int result_timer_;
    const int RESULT_DISPLAY_FRAMES_;
    const int AUTO_RESTART_FRAMES_;
    
    // Independent parallel monitoring variables
    bool liveness_passed_;
    bool face_recognized_;
    std::string recognized_person_;
    double last_recognition_score_;
};

int main() {
    // Keep your exact original main loop initialization logic 
    AttendancePipeline pipeline;
    cv::VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cerr << "Failed to open camera" << std::endl;
        return 1;
    }

    std::cout << "\n=== ATTENDANCE PIPELINE ===" << std::endl;
    std::cout << "SPACE - Start / restart anytime" << std::endl;
    std::cout << "R     - Register new face" << std::endl;
    std::cout << "Q     - Quit" << std::endl;

    bool register_mode = false;
    std::string register_name;

    cv::Mat frame;
    while (cap.read(frame)) {
        cv::flip(frame, frame, 1);

        if (register_mode) {
            cv::putText(frame, "REGISTER MODE: " + register_name, {30, 50}, cv::FONT_HERSHEY_SIMPLEX, 0.8, cv::Scalar(0, 255, 255), 2);
            cv::putText(frame, "Press ENTER to capture", {30, 90}, cv::FONT_HERSHEY_SIMPLEX, 0.6, cv::Scalar(255, 255, 255), 1);
        } else {
            auto [display, result, state] = pipeline.process_frame(frame);
            frame = display;
        }

        cv::imshow("Attendance System", frame);

        int key = cv::waitKey(1) & 0xFF;
        if (key == 'q' || key == 'Q') break;
        else if (key == ' ' && !register_mode) {
            pipeline.start();
        } else if ((key == 'r' || key == 'R') && !register_mode) {
            std::cout << "Enter person name/ID: ";
            std::getline(std::cin, register_name);
            register_mode = true;
        } else if (key == 13 && register_mode) { 
            auto [ok, msg] = pipeline.register_face(frame, register_name);
            std::cout << msg << std::endl;
            register_mode = false;
            register_name = "";
        }
    }

    cap.release();
    cv::destroyAllWindows();
    return 0;
}